fix: verify length and hashes of fetched bytes before parsing

joshuagl opened the pull request:
Fixes #233
Release Notes: <!-- What comments/remarks should we include in the release notes for this change? -->
Check hashes and length of downloaded snapshot and targets metadata before parsing.

**Types of changes**:
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Description of the changes being introduced by the pull request**:
Better match the specification by performing checks against length and hashes on the downloaded bytes before parsing the bytes into a FileMeta.

**Please verify and check that the pull request fulfills the following requirements**:

- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

mnm678 commented on util/util.go:
```diff
@@ -86,6 +86,33 @@ func FileMetaEqual(actual data.FileMeta, expected data.FileMeta) error {
+		return ErrWrongLength{length, flen}
+	}
+
+	if len(hashes) != 0 {
```
Should this error if there are no hashes?

joshuagl replied:
I don't think so, length and hashes are both optional for [`METAFILES`](https://theupdateframework.github.io/specification/latest/#metafiles).

znewman01 commented on util/util.go:
```diff
@@ -86,6 +86,33 @@ func FileMetaEqual(actual data.FileMeta, expected data.FileMeta) error {
+		return ErrWrongLength{length, flen}
+	}
+
+	if len(hashes) != 0 {
```
Why do we need the `!= 0` check? (This is probably my Go inexperience showing)

rdimitrov replied:
That's right, the `len` check can be omitted since it will skip the `for` loop if there are no `hashes` entries 👍

joshuagl replied:
Well spotted both, thanks! This was my Go inexperience showing :-) Removed the `len()` check in [39b9b0f](https://github.com/theupdateframework/go-tuf/pull/325/commits/39b9b0f1c495d2be5ff14317f0248dce5327e05d)

rdimitrov reviewed (changes requested):
I'd suggest we add a few simple unit tests for `BytesMatchLenAndHashes` and `VersionEqual` in `util_tests.go`, so we know in case they break at some point 👍 Other than that, it's great 👍 🙏

joshuagl commented:
I've add some tests in af46f5d and removed the unnecessary `len()` check in 39b9b0f. I've also rebased on the latest master branch and force pushed.

PTAL @mnm678 @znewman01  @rdimitrov 🙏

rdimitrov reviewed (approved):
Thanks! 👍 LGTM! 💯
